src/main.cpp

```diff
@@ -138,77 +138,37 @@ void DDWAF::dispose(const Napi::CallbackInfo& info) {
-    Napi::Error::New(env, "Could not update rule data on a disposed WAF").ThrowAsJavaScriptException();
-    return;
-  }
-  if (info.Length() < 1) {
-    Napi::Error::New(env, "Wrong number of arguments, expected 1").ThrowAsJavaScriptException();
+
+  size_t arg_len = info.Length();
+  if (arg_len < 1) {
+    Napi::Error::New(env, "Wrong number of arguments, expected at least 1").ThrowAsJavaScriptException();
     return;
   }
-  if (!info[0].IsArray()) {
-    Napi::TypeError::New(env, "First argument must be an array").ThrowAsJavaScriptException();
+  if (!info[0].IsObject()) {
+    Napi::TypeError::New(env, "First argument must be an object").ThrowAsJavaScriptException();
     return;
   }
 
-  ddwaf_object data;
-  to_ddwaf_object(&data, env, info[0], 0, false);
-
-  DDWAF_RET_CODE code = ddwaf_update_rule_data(this->_handle, &data);
-  ddwaf_object_free(&data);
-
-  switch (code) {
-    case DDWAF_ERR_INTERNAL:
-      Napi::Error::New(env, "Internal error").ThrowAsJavaScriptException();
-      break;
-    case DDWAF_ERR_INVALID_OBJECT:
-      Napi::Error::New(env, "Invalid ddwaf object").ThrowAsJavaScriptException();
-      break;
-    case DDWAF_ERR_INVALID_ARGUMENT:
-      Napi::Error::New(env, "Invalid arguments").ThrowAsJavaScriptException();
-      break;
-    default:
-      break;
-  }
-}
-
-void DDWAF::toggleRules(const Napi::CallbackInfo& info) {
-  mlog("Toggling rules on DDWAF");
-  Napi::Env env = info.Env();
   if (this->_disposed) {
-    Napi::Error::New(env, "Could not toggle rule on a disposed WAF").ThrowAsJavaScriptException();
-    return;
-  }
-  if (info.Length() < 1) {
-    Napi::Error::New(env, "Wrong number of arguments, expected 1").ThrowAsJavaScriptException();
-    return;
-  }
-  if (!info[0].IsObject()) {
-    Napi::TypeError::New(env, "First argument must be an object").ThrowAsJavaScriptException();
+    Napi::Error::New(env, "Could not update a WAF disposed instance").ThrowAsJavaScriptException();
     return;
   }
 
-  ddwaf_object rulesToggleMap;
-  to_ddwaf_object(&rulesToggleMap, env, info[0], 0, false);
+  ddwaf_object update;
+  mlog("building rules update");
+  to_ddwaf_object(&update, env, info[0], 0, false);
 
-  DDWAF_RET_CODE code = ddwaf_toggle_rules(this->_handle, &rulesToggleMap);
-  ddwaf_object_free(&rulesToggleMap);
+  mlog("Update DDWAF instance");
+  ddwaf_ruleset_info_free(&this->_ruleset_info);
+  ddwaf_handle updated_handle = ddwaf_update(this->_handle, &update, &this->_ruleset_info);
+  ddwaf_object_free(&update);
 
-  switch (code) {
-    case DDWAF_ERR_INTERNAL:
-      Napi::Error::New(env, "Internal error").ThrowAsJavaScriptException();
-      break;
-    case DDWAF_ERR_INVALID_OBJECT:
-      Napi::Error::New(env, "Invalid ddwaf object").ThrowAsJavaScriptException();
-      break;
-    case DDWAF_ERR_INVALID_ARGUMENT:
-      Napi::Error::New(env, "Invalid arguments").ThrowAsJavaScriptException();
-      break;
-    default:
-      break;
+  if (updated_handle != NULL) {
```

Comment:
If the handle is null then an error happened, shouldn't we throw if it is ?

Reply:
> The ddwaf_update function returns a new ddwaf_handle which will be a valid pointer if the update succeeded, or NULL if there was nothing to update or there was an error.

So, at this point is not sure that an error has happened or there is nothing to be updated, we should check `ruleset_info` for that. Does it should be checked in the bindings or in the tracer? That's the question...

Reply:
As discussed, an error when updating or a nothing-to-update situation will result in an exception to be thrown. Then it can be caught in JS realm to be logged properly.